src/datasets/tasks/base.py

```diff
@@ -0,0 +1,30 @@
+import abc
+from dataclasses import dataclass
+from typing import Dict
+
+from ..features import Features
+
+
+@dataclass
+class TaskTemplate(abc.ABC):
+    task: str
+    input_schema: Features
+    label_schema: Features
+
+    @property
+    def features(self) -> Features:
+        return Features(**self.input_schema, **self.label_schema)
+
+    @property
+    @abc.abstractmethod
+    def column_mapping(self) -> Dict[str, str]:
+        return NotImplemented
+
+    @classmethod
+    @abc.abstractmethod
+    def from_dict(cls, template_dict: dict) -> "TaskTemplate":
+        return NotImplemented
+
+    @abc.abstractmethod
+    def __hash__(self):
+        raise NotImplementedError
```

Comment:
I'd rather mark the dataclass as "frozen": it would generate the `__hash__` method for us (see https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass, the paragraph about `hash`)

I think it makes sense to mark a TaskTemplate as immutable, it should only be specified once when instantiating the `DatasetInfo` object.

```python
@dataclass(frozen=True)
class TaskTemplate(abc.ABC):
    # [...]
```

Reply:
i've had a stab at this and am running into an issue that you might know the answer to :)

first, if we mark `TaskTemplate` as immutable with `frozen=True` than any subclass must also be marked as immutable (we get a `TypeError` otherwise)

with that in mind, let's modify our `TextClassification` template as follows:

```python
@dataclass(frozen=True)
class TextClassification(TaskTemplate):
    task = "text-classification"
    input_schema = Features({"text": Value("string")})
    label_schema = Features({"labels": ClassLabel})
    labels: List[str]
    text_column: str
    label_column: str

    def __init__(
        self,
        labels,
        text_column="text",
        label_column="labels",
    ):
        assert sorted(set(labels)) == sorted(labels), "Labels must be unique"
        object.__setattr__(self, "labels", sorted(labels))
        object.__setattr__(self, "text_column", text_column)
        object.__setattr__(self, "label_column", label_column)
        self.label_schema["labels"] = ClassLabel(names=self.labels)
        object.__setattr__(self, "label2id", {label: idx for idx, label in enumerate(self.labels)})
        object.__setattr__(self, "id2label", {idx: label for label, idx in self.label2id.items()})
```

note that the use of `object.__setattr__` is needed to initialize the fields (see [docs](https://docs.python.org/3/library/dataclasses.html#frozen-instances)).

the problem i'm running into is that `labels` is of type `list` and thus not hashable by default. i can exclude it from the comparison methods like `__eq__` by defining:

```python
from dataclasses import field 

@dataclass(frozen=True)
class TextClassification(TaskTemplate):
    task = "text-classification"
    input_schema = Features({"text": Value("string")})
    label_schema = Features({"labels": ClassLabel})
    labels: List[str] = field(compare=False)
    text_column: str
    label_column: str
```
but this isn't ideal because now the hash for two `TextClassifcation` instances with different labels, but identical column names is the same:

```python
tc1 = TextClassification(labels=["pos", "neg"])
tc2 = TextClassification(labels=["pos", "neg", "neutral"])
# returns True
hash(tc1) == hash(tc2)
```

do you happen to know of an elegant way to incorporate the information in `labels` into the dataclass' `__hash__` function? in my original implementation i used `frozenset(self.labels)` but could not find a nice way to generate this behaviour automagically from the dataclass.

Reply:
Maybe you can use a tuple instead of a list ?

Reply:
Maybe you can use a tuple instead of a list ?

Reply:
resolved by @SBrandeis' great idea to cast `labels` to a tuple 🚀